src/tool/Plugins.cpp

```diff
@@ -0,0 +1,72 @@
+//
+// This is a derivative work. originally part of the LLVM Project.
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// Copyright (c) 2023 Vinnie Falco (vinnie.falco@gmail.com)
+//
+// Official repository: https://github.com/cppalliance/mrdox
+//
+
+#include <mrdox/Plugin.hpp>
+#include <mrdox/Version.hpp>
+#include <mrdox/Support/Path.hpp>
+
+#include "lib/Support/GeneratorsImpl.hpp"
+
+#include <llvm/Support/DynamicLibrary.h>
+
+namespace clang {
+namespace mrdox {
+
+class PluginEnvironmentImpl final : public PluginEnvironment
+{
+public:
+    void
+    addGenerator(
+        std::unique_ptr<Generator> generator) override
+    {
+        getGeneratorsImpl().insert(std::move(generator));
+    }
+};
+
+Error loadOnePlugin(
+    std::string pth,
+    PluginEnvironment & env)
+{
+    if (pth.ends_with(".so") || pth.ends_with(".dll"))
+    {
+        std::string errMsg;
+        auto res = llvm::sys::DynamicLibrary::getPermanentLibrary(pth.c_str(), &errMsg);
+        if (!res.isValid())
+            return formatError("Couldn't load {}, because '{}'", pth, errMsg);
+
+        auto f = res.getAddressOfSymbol("MrDoxMain");
+        if (f == nullptr)
+            return formatError("{}, doesn't export MrDoxMain symbol", pth);
+
+        auto func = reinterpret_cast<decltype(&MrDoxMain)>(f);
+        if (!func(projectVersionMajor, projectVersionMinor, env))
```

Comment:
This is not right. What if the plugin failed to load because a resource file it needs is missing? Should it return true or false? Having the plugin return the bool is too clever and needlessly couples the host and the plugin. Just return void from MrDoxMain and let the plugin do its own version checking and reporting.

Reply:
It uses `report::error`.

Reply:
This is your code:
```
        if (!func(projectVersionMajor, projectVersionMinor, env))
            return formatError("Couldn't load {} - version mismatch.", pth);
```

Now lets say MrDoxMain tries to load a resource file and it fails. But the version is ok. Should MrDoxMain return "true", or "false?" If it returns false then the error message will be incorrect. It will say "version mismatch" but it is really a "missing file" error.

If it returns true then the reader will infer that the plugin loaded successfully even though it could not carry out its installation.

Do you understand my point? Giving MrDoxMain a return value was not a good design choice. Do you remember the signature I gave you?

![image](https://github.com/cppalliance/mrdox/assets/1503976/f281aaa6-dd53-4972-9dac-18a66cf1acce)

It was like that for a reason.